#Imports

In [20]:
!source venv/bin/activate

# this is for our data manipulations
import pandas as pd
import numpy as np

# this is for our visualizations
import matplotlib.pyplot as plt 
import seaborn as sns

plt.rcParams['figure.figsize'] = [8,5]
plt.rcParams['font.size'] =14
plt.rcParams['font.weight']= 'bold'

plt.style.use('https://github.com/dhaitz/matplotlib-stylesheets/raw/master/pitayasmoothie-light.mplstyle')

zsh:source:1: no such file or directory: venv/bin/activate


###Charge dataset
>So here, we can fastly find our $m$,$n$ values  
>$n$=columns, $m$=rows  

In [6]:
# import the dataset
#path ==> 'data/'

path = '../data/'
n_rows, n_cols = df.shape

df = pd.read_csv(path+'insurance.csv')
print(f"Dataset columns : {n_cols}\nDataset rows : {n_rows}")
print("")

# looking into top rows and columns
df.head()

Dataset columns : 7
Dataset rows : 1338



age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

>$n$=7, $m$=1338  
>With $m$ our number of training examples, and $n$ our number of independant variables  

>Here :  
>- the features $x$ are : ***age, sex, bmi, children, smoker, region***  
>- the target variable $y$ is : ***charges***  

>That means we need to fit multiple linear regression  
>In this project, we will use teta $(\theta)$ as a vector of $w,b$
>
$$
\theta = w,b
$$
>To compute the medical cost for a ***i*** person :
$$
h_\theta(x_i)
$$  
>So the hypothesis function for our dataset :  
$$
h_\theta(x_i) = \theta_0 + \theta_1 \cdot \text{age} + \theta_2 \cdot \text{sex} + \theta_3 \cdot \text{bmi} + \theta_4 \cdot \text{children} + \theta_5 \cdot \text{smoker} + \theta_6 \cdot \text{region}
$$

>So,  
>If $i = 1$,  
$$
h_\theta(x_1) = \theta_0 + \theta_1 \cdot 19 + \theta_2 \cdot \text{female} + \theta_3 \cdot 27.900 + \theta_4 \cdot 1 + \theta_5 \cdot \text{yes} + \theta_6 \cdot \text{southwest}
$$
$$
y1 = 168884.92400
$$

>Because we have : 
<p style="text-align:center;">
  <img src="../images/01_x1.png">
</p>  

>$\theta_0$ = constant (bias/intercept), in french : biais/ordonnée à l'origine  
>In fact, it's when all features = 0  
>$\theta_1$ = coefficient of the 1st feature  
>$\theta_2$ = coefficient of the 2nd feature  
>...

>It's determining the ***height*** of a feature $x$ on the ***model prediction***  
>Keep in mind that the best value of $\theta$ is the one who minimize the most the errors between predictions & true values.
>/!\ Index start from 0, so the first element $i$=1 is index 0



>So our goal now will be to find best value of $\theta$

#Cost function

>The cost function measaure the errors between predicted and real values in our model.  
$$
J(\theta) = \frac{1}{m} \sum_{i=1}^{m} (h_\theta(x_i) - y_i)^2
$$
>It's represent the mean of the squared errors

>To implement the linear regression, take training example add an extra column that is  $X_{0}$ feature  
[\
$X$ = 
\begin{pmatrix}
x_{10} & x_{11} & x_{12} & \dots & x_{1n} \\
x_{20} & x_{21} & x_{22} & \dots & x_{2n} \\
x_{m0} & x_{m1} & x_{m2} & \dots & x_{mn}
\end{pmatrix}_{(m,n+1)}
\]

Now, we rewrite the least square cost function in matrix form :

$$
J(\theta) = \frac{1}{m} \left( X\theta - y \right)^T \left( X\theta - y \right)
$$



#J derivative

>To minimize our cost function and also obtain optimal values for $\theta$, we need to compute the derivative of J($\theta$)
$$
\frac{\partial J(\theta)}{\partial \theta_j}J(\theta)
$$
